src/main/scala/akka/persistence/jdbc/query/scaladsl/JdbcReadJournal.scala

```diff
@@ -123,36 +123,53 @@ class JdbcReadJournal(config: Config, configPath: String)(implicit val system: E
     adapter.fromJournal(repr.payload, repr.manifest).events.map(repr.withPayload)
   }
 
-  private def currentJournalEventsByPersistenceId(persistenceId: String, fromSequenceNr: Long, toSequenceNr: Long): Source[PersistentRepr, NotUsed] =
-    readJournalDao.messages(persistenceId, fromSequenceNr, toSequenceNr, Long.MaxValue)
+  private def currentJournalEventsByPersistenceId(persistenceId: String, fromSequenceNr: Long, toSequenceNr: Long, max: Long): Source[PersistentRepr, NotUsed] =
+    readJournalDao.messages(persistenceId, fromSequenceNr, toSequenceNr, max)
       .mapAsync(1)(deserializedRepr => Future.fromTry(deserializedRepr))
 
   override def currentEventsByPersistenceId(persistenceId: String, fromSequenceNr: Long, toSequenceNr: Long): Source[EventEnvelope, NotUsed] =
-    currentJournalEventsByPersistenceId(persistenceId, fromSequenceNr, toSequenceNr)
+    currentJournalEventsByPersistenceId(persistenceId, fromSequenceNr, toSequenceNr, Long.MaxValue)
       .mapConcat(adaptEvents)
       .map(repr => EventEnvelope(Sequence(repr.sequenceNr), repr.persistenceId, repr.sequenceNr, repr.payload))
 
-  override def eventsByPersistenceId(persistenceId: String, fromSequenceNr: Long, toSequenceNr: Long): Source[EventEnvelope, NotUsed] =
-    Source.unfoldAsync[Long, Seq[EventEnvelope]](Math.max(1, fromSequenceNr)) { (from: Long) =>
-      def nextFromSeqNr(xs: Seq[EventEnvelope]): Long = {
-        if (xs.isEmpty) from else xs.map(_.sequenceNr).max + 1
-      }
-      from match {
-        case x if x > toSequenceNr => Future.successful(None)
-        case _ =>
-          delaySource
-            .flatMapConcat { _ =>
-              currentJournalEventsByPersistenceId(persistenceId, from, toSequenceNr)
-                .take(readJournalConfig.maxBufferSize)
-            }
-            .mapConcat(adaptEvents)
-            .map(repr => EventEnvelope(Sequence(repr.sequenceNr), repr.persistenceId, repr.sequenceNr, repr.payload))
-            .runWith(Sink.seq).map { xs =>
-              val newFromSeqNr = nextFromSeqNr(xs)
-              Some((newFromSeqNr, xs))
+  override def eventsByPersistenceId(persistenceId: String, fromSequenceNr: Long, toSequenceNr: Long): Source[EventEnvelope, NotUsed] = {
+    import JdbcReadJournal._
+    val batchSize = readJournalConfig.maxBufferSize
+
+    Source.unfoldAsync[(Long, FlowControl), Seq[PersistentRepr]]((Math.max(1, fromSequenceNr), Continue)) {
+      case (from, control) =>
+        def retrieveNextBatch(): Future[Option[((Long, FlowControl), Seq[PersistentRepr])]] = {
+          for {
+            xs <- currentJournalEventsByPersistenceId(persistenceId, from, toSequenceNr, batchSize).runWith(Sink.seq)
+          } yield {
+            val hasMoreEvents = xs.size == batchSize
+            // Events are ordered by sequence number, therefore the last one is the largest
+            val hasLastEvent = xs.lastOption.exists(last => last.sequenceNr >= toSequenceNr)
```

Comment:
We could have an optimization here: 

```scala
val lastSeqNrInBatch: Option[PersistentRepr] = xs.lastOption.map(_.sequenceNr)
val hasLastEvent = lastSeqNrInBatch.exists(_ >= toSequenceNr)

... then later

val nextFrom: Long = 
  lastSeqNrInBatch match {
    // Continue querying from the last sequence number (the events are ordered)
    case Some(lastSeqNr) => lastSeqNr + 1
    case None =>  from
  }
```

As it is now, we are traversing `xs` twice. It's a tiny optimization, but helpful when `xs` is large.

Reply:
Implemented this.
I am not sure how much is really matters since `xs` should be a vector, and we only look at the last element, but anyways this is could be slightly better.